File: packages/cli-python/src/spec_init/commands/add_skill.py
"""spec-init add-skill <name>.

Copies a skill definition from the payload's skills/ catalog into the
project's .claude/skills/ directory. Stage 5 populates the catalog.
"""

from __future__ import annotations

import shutil
import sys
from pathlib import Path

from ..args import CliOptions, UsageError
from ..payload import resolve_payload_dir


def run_add_skill(opts: CliOptions) -> None:
    """Copy a skill from the payload catalog. Raises UsageError on refusals."""
    if not opts.positional:
        raise UsageError("add-skill requires a name: spec-init add-skill <name>")
    name = opts.positional[0]

    payload_root = resolve_payload_dir().parent
    catalog = payload_root / "skills"
    available = sorted(p.name for p in catalog.iterdir()) if catalog.exists() else []

    if not available:
        sys.stderr.write(
            "spec-init: no skills available in this payload (populated in Stage 5).\n"
        )
        raise UsageError(f"add-skill: unknown skill '{name}'")

    source = catalog / name
    if not source.exists():
        raise UsageError(
            f"add-skill: unknown skill '{name}' (available: {', '.join(available)})"
        )

    dest = Path.cwd() / ".claude" / "skills" / name
    if opts.dry_run:
        sys.stdout.write(f"[dry-run] would copy {source} -> {dest}\n")
        return

    dest.parent.mkdir(parents=True, exist_ok=True)
    shutil.copytree(source, dest)
    sys.stdout.write(f"[OK] installed skill: {name}\n")

File: packages/cli-node/src/commands/add-skill.ts
// spec-init add-skill <name>: copy a skill from the payload catalog into .claude/skills/.

import { cpSync, existsSync, mkdirSync, readdirSync } from 'node:fs';
import { join, resolve } from 'node:path';
import { UsageError } from '../args.js';
import type { CliOptions } from '../args.js';
import { resolvePayloadDir } from '../payload.js';

/** Run `add-skill`. Throws UsageError on missing name or unknown skill. */
export function runAddSkill(opts: CliOptions): void {
  const [name] = opts.positional;
  if (!name) throw new UsageError('add-skill requires a name: spec-init add-skill <name>');

  const payloadRoot = resolve(resolvePayloadDir(), '..');
  const catalog = join(payloadRoot, 'skills');
  const available = existsSync(catalog) ? readdirSync(catalog) : [];

  if (available.length === 0) {
    process.stderr.write(
      'spec-init: no skills available in this payload (populated in Stage 5).\n',
    );
    throw new UsageError(`add-skill: unknown skill '${name}'`);
  }

  const source = join(catalog, name);
  if (!existsSync(source)) {
    throw new UsageError(`add-skill: unknown skill '${name}' (available: ${available.join(', ')})`);
  }

  const projectRoot = resolve(process.cwd());
  const dest = join(projectRoot, '.claude', 'skills', name);
  if (opts.dryRun) {
    process.stdout.write(`[dry-run] would copy ${source} -> ${dest}\n`);
    return;
  }

  mkdirSync(dest, { recursive: true });
  cpSync(source, dest, { recursive: true });
  process.stdout.write(`[OK] installed skill: ${name}\n`);
}
